### ISSUE TITLE
Context Site should be visible through Factory.GetSite()

### ISSUE THREAD

**Issue by dharnitski** [sergeyshushlyapin/Sitecore.FakeDb#123]
Current way to work with site affects only Context but other APIs like Factory.GetSite() are not aware about site we use.

Updated test case for Site should look as follows:

```csharp
    [Fact]
    public void HowToSwitchContextSite()
    {
      // Create a fake Site Context and configure the required parameters.
      // Please note that there is no registration in the App.config file required.
      var fakeSite = new Sitecore.FakeDb.Sites.FakeSiteContext(
        new Sitecore.Collections.StringDictionary
          {
            { "name", "website" }, { "database", "web" }
          });

      // switch the context site
      using (new Sitecore.FakeDb.Sites.FakeSiteContextSwitcher(fakeSite))
      {
        Xunit.Assert.Equal("website", Sitecore.Context.Site.Name);
        Xunit.Assert.Equal("web", Sitecore.Context.Site.Database.Name);
        Xunit.Assert.Equal("website", Factory.GetSite("website").Name);
      }
    }
```

**Comment by sergeyshushlyapin** [comment]
Hi @dharnitski,
Thank you for your interest in FakeDb!

> Current way to work with site affects only Context

Yes, and that is what the `Sitecore.Sites.SiteContextSwitcher` class [supposed](http://ctor.io/correctly-switching-sitecore-contextes/) to do.

> but other APIs like Factory.GetSite() are not aware about site we use

That is how `Sitecore.Sites.SiteContextFactory` works. It knows nothing about the `Context.Site` and just read sites from the config. Everything works *as is* here, no FakeDb involved and I believe that is good.

You probably might be interesting in substituting the factory itself, but unfortunately from what I see it's impossible. It's just static, there is no way to inject a *mocked provider* or something like that. If you need the factory to return some sites, registering them in the config might be an option.

**Comment by dharnitski** [comment]
Hi Sergey,

Sorry for confusion with this issue. I wanted to create it only as documentation for my Pull request.
Please check https://github.com/sergeyshushlyapin/Sitecore.FakeDb/pull/124

These two methods can be used for mocking Sites:

```csharp
SiteContextFactory.Reset();
SiteContextFactory.Sites.Add(); 
```

SiteContextFactory works as cache. You can temporary fill in with what you need for test and clean using Reset() when you are done. It will load sites from configured provider when next caller use Factory.

**Comment by sergeyshushlyapin** [comment]
There are few concerns preventing me from accepting the PR. Analysing the **existing** Sitecore API behaviour:
  - *SiteContextSwitcher* does not use *SiteContextFactory*
  - *SiteContextFactory* does not use *SiteContextSwitcher*

Why should FakeDb change this and *"register"* a new site in the factory when the fake switcher is used? This leads to the inconsistent behaviour where there are two switcher available at the same time. The default one, which just *switches* the context site, and the new one that *switches and registers* the site it in the factory.

IMO, these concepts should be kept separate. One can easily add sites to the factory (if needed) as you [showed](https://github.com/dharnitski/Sitecore.FakeDb/blob/d3e70a78bcbe02e3a6dc938e1f63c65d1511dace/src/Sitecore.FakeDb.Tests/Sites/FakeSiteContextSwitcherTest.cs) in the PR.

Does it may sense?
//cc: @pveller

**Comment by dharnitski** [comment]
Hi Sergey,

That makes sense. After your comment I better see the intent of this project.

Although I wondering how we can help developers who need to mock Sites. I spent a few hours before came to this workaround based on *SiteContextFactory*. 

Could you at least add something to Wiki  https://github.com/sergeyshushlyapin/Sitecore.FakeDb/wiki/Sites page or I can add some Unit Tests for manual Adding Sites and Reseting *SiteContextFactory* execution without introducing new *SiteContextSwitcher*?

**Comment by pveller** [comment]
I will add to what @sergeyshushlyapin said. FakeDb is all about enabling Sitecore APIs and functionality as-is without a database. 

We sometimes have to build and mimic a certain piece of Sitecore runtime when we can't just isolate or it's not practical to replicate/recreate underlying data structures. Standard Values is a good example. We built our own provider to add support for standard values (as in "standard values the concept") without requiring that fake templates have fake `__Standard values` items. Another example is auto-creation of fake field items under fake template items. In doing so we replicated what Sitecore does - the effect that users/developers see and sometimes the underlying data structures - but we haven't altered it or haven't added *unexpected* side effects. 

This is probably one of the guiding principle that FakeDb has - we try not to produce unexpected side effects or otherwise alter Sitecore behavior. I will say though that if there was a very helpful shortcut that everybody needs/does in their tests - FakeDb will be happy to incorporate it. We did this, for example, with the FakeLinkDatabase. It's not *switching* in Sitecore but we made it so to provide isolation for tests. I originally built it in my test project and then later ported to FakeDb. 

We basically need to decide if we like the new site context switcher that also registers a site temporarily with the factory enough to add it to the product. I also wonder if we need to worry about the new site providers in 8. Not every Sitecore build sources site configuration via XML config patches and Factory, right?

**Comment by sergeyshushlyapin** [comment]
I agree with @pveller. Standard Values and auto-generated fields are excellent examples of features that allow developers to write much less code comparing to the native Sitecore API. I'm still in doubts if the new Site Switcher may have similar impact (taking into account that **there is** SiteContextSwitcher everyone is aware about)...

> I also wonder if we need to worry about the new site providers in 8

Are you talking about `Sitecore.Sites.SiteProvider`? From what I see it can be successfully used in tests if necessary:
```csharp
[Theory, DefaultSubstituteAutoData]
public void ShouldGetFakeSite(SiteProvider siteProvider, Site site)
{
  siteProvider.Name.Returns("fake");
  siteProvider.GetSite("mysite").Returns(site);
  SiteManager.Providers.Add(siteProvider);

  using (new SiteProviderSwitcher("fake"))
  {
    SiteManager.GetSite("mysite").Should().BeSameAs(site);
  }
}
```

**Comment by pveller** [comment]
Right, I was trying to say that some will need their fake site to be readable from the `Factory` and some others will want their fake site to make to the `SiteManager`. I haven't used theses APIs a lot so maybe I am saying something that doesn't make sense :)

**Comment by sergeyshushlyapin** [comment]
Good news, gents. In order to configure the SiteContextFactory, one can mock the existing SiteProvider. Look at this working example:
```csharp
[Theory, DefaultSubstituteAutoData]
public void ShouldGetFakeSite(SiteProvider siteProvider)
{
  try
  {
    var site = new Site("mysite", new StringDictionary { { "name", "mysite" } });

    siteProvider.Name.Returns("fake");
    siteProvider.GetSites().Returns(new SiteCollection { site });
    SiteManager.Providers.Add(siteProvider);

    using (new SiteProviderSwitcher("fake"))
    {
      SiteContextFactory.GetSiteInfo("mysite").Should().NotBeNull();
      SiteContextFactory.GetSiteNames().Should().Contain(site.Name);
    }
  }
  finally
  {
    SiteContextFactory.Reset();
  }
}
```

That is a raw poc. Moving the `try-catch` block to smth. like xUnit's [BeforeAfterTestAttribute](http://www.bricelam.net/2012/04/xunitnet-extensibility.html) and registering the fake provider in the config can make the test half as large as now. I'll continue on this.

**Comment by sergeyshushlyapin** [comment]
Just wondering if that is something new (SiteProvider) or I've simply overlooked it in the fullness of time.

**Comment by pveller** [comment]
I think one could always override the site resolver or something (that's how folks would do microsites frameworks that would allow adding new sites with 0 configuration change - no need to patch `<sites>`) but `SiteProvider` concept is new to 8. Unless I too have overlooked it :)

**Comment by sergeyshushlyapin** [comment]
Unfortunately my experiments with Site Providers didn't work out. As @dharnitski mentioned above, the problem lies in the static `searchTable` dictionary that caches sites once they're returned from the provider. The only way to bypass this is to *reset* the factory "back and forth" all the time (actually, that is what @dharnitski has implemented in the *FakeSiteContextSwitcher* constructor and `Dispose()` method).

>Although I wondering how we can help developers who need to mock Sites. I spent a few hours before came to this workaround based on SiteContextFactory.

You're right. That is a trade off between the *correct API* and API that allows developers not to think about the necessity to *reset* the factory manually before and after each operation.

I'm appreciate your efforts and intention to contribute to the FakeDb project. I see your desire to simplify the site configuration mechanism, and I almost convinced myself that the suggested *FakeSiteContextSwitcher* can solve the problem, but then I found the following interesting test in the PR code:
```csharp
[Fact]
public void ShouldPreserveExistingSites()
{
    // arrange
    var existingSite = new FakeSiteContext("existing");
    SiteContextFactory.Sites.Add(existingSite.SiteInfo);

    var site = new FakeSiteContext("mywebsite");

    // act
    using (new FakeSiteContextSwitcher(site))
    {
        // assert
        Context.Site.Name.Should().Be("mywebsite");
        Factory.GetSite("existing").Should().NotBeNull();
    }

    Context.Site.Should().BeNull();
    Factory.GetSite("mywebsite").Should().BeNull();
    Factory.GetSite("existing").Should().NotBeNull();
}
```

Do you see what I see? ;)

- In the *arrange* section, the "existing" site is added to the *global* Sites collection:
`SiteContextFactory.Sites.Add(existingSite.SiteInfo);`

- In the *assert* section (the last line of the test), there is the assertion that states that the "existing" site is **still in place**:
`    Factory.GetSite("existing").Should().NotBeNull();`

It means that the **test leaves garbage** after itself. That might be OK for a while, but at some point one **will** end up with random failures.

So, the only option I see here is to add and clean up the sites **consciously**:
```csharp
internal class ResetSiteContextFactoryAttribute : BeforeAfterTestAttribute
{
  public override void After(MethodInfo methodUnderTest)
  {
    SiteContextFactory.Reset();
  }
}

[Fact]
[ResetSiteContextFactory]
public void ShouldPreserveExistingSites()
{
  // ...

  // the site is here for now but it'll be reset by xunit later on
  Factory.GetSite("existing").Should().NotBeNull();
}
```

**Comment by dharnitski** [comment]
Hi, @sergeyshushlyapin  
Let me explain the idea behind keeping existing sites in *SiteContext*. 

As you mentioned, developer can create some sites in *web.config* or use any other way to cover *happy path* configuration. 

These two lines of code in my test is just emulation for something that should be in *web.config* (I should add comment clarifying that):

```csharp
var existingSite = new FakeSiteContext("existing");
    SiteContextFactory.Sites.Add(existingSite.SiteInfo);
```

*FakeSiteContextSwitcher* has special code to persist sites configuration that was available before Switcher object is created and restore state when object is disposed.

If you feel that support of "happy path" should not be here we can drop it. I built it because we already rely on sites in .config in our tests and I did not want to break existing code.

It also make perfect sense to express what we are doing explicitly. You are right that *SiteContextSwitcher* has its own purpose. Maybe we can introduce another class that will clearly shows by its name that it is not only about Site Context. Something like SitesMock (or maybe you can propose better name) will setup all Sites related API for test. 

We can overload constructors so we can cover not only current Context Site but can also add secondary sites to *SiteContextFactory*. Not primary sites are be used in our code. One of the scenario is  generate links to another tenant. Another is redirect to regional site based in IP detection.

```csharp
public class SitesMock: IDisposable
    {
        public SitesMock(IEnumerable<KeyValuePair<string, string>> site)
        {     }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="currentSite"></param>
        /// <param name="secondarySites"></param>
        public SitesMock(IEnumerable<KeyValuePair<string, string>> currentSite, IEnumerable<IEnumerable<KeyValuePair<string, string>>> secondarySites)
        {
            //code to set up Sites environment
        }

        public void Dispose()
        {
            //code to set cleanup Sites environment
        }
    }
```

```csharp
 public void ShouldTestSetupCleanup()
        {
            //make sure that we create these sites from scratch 
            Context.Site.Should().BeNull();
            Factory.GetSite("website").Should().BeNull();


            //arrange
            var currentSite = new StringDictionary
                {
                    {"name", "website"}
                };

            var secondatySites = new List<StringDictionary>
                {
                    new StringDictionary
                        {
                            {"name", "secondsite"}
                        }
                };

            //act
            using (new SitesMock(currentSite, secondatySites))
            {
                //assert environment setup 
                Context.Site.Name.Should().Be("website");
                Factory.GetSite("website").Should().NotBeNull();
                Factory.GetSite("secondsite").Should().NotBeNull();
                //site provider should not be tested here because it should not be used directly from business logic code 
            }

            //assert environment clean up 
            Context.Site.Should().BeNull();
            Factory.GetSite("website").Should().BeNull();
            Factory.GetSite("secondsite").Should().BeNull();
        }
```

@pveller @sergeyshushlyapin let me know what do you think and I will be happy to send another pull request.

**Comment by sergeyshushlyapin** [comment]
> As you mentioned, developer can create some sites in web.config or use any other way to cover happy path configuration.
These two lines of code in my test is just emulation for something that should be in web.config (I should add comment clarifying that):

No, everything is pretty clear here. The code perfectly demonstrates **how easy one can add sites** in code and then forget to clean them up properly! That was my biggest concern I wanted to highlight ;)

From the other side, *FakeSiteContextSwitcher* takes care about both *types* of sites (`SiteContext` and `SiteInfo`) which allows to replace the following code:
```csharp
var site = SiteInfo.Create(new StringDictionary { { "name", "website" } });
SiteContextFactory.Sites.Add(site);

using (new SiteContextSwitcher(new SiteContext(site)))
{
  // ...
}
```

with this one:
```csharp
using (new FakeSiteContextSwitcher(new FakeSiteContext("mywebsite")))
{
  // ...
}
```

Now I found this simplification useful. I'll merge it right away. Thanks for your patience, @dharnitski :)

**Comment by sergeyshushlyapin** [comment]
Now available as [v0.38.0](https://www.nuget.org/packages/Sitecore.FakeDb/).
